### PR TITLE
chore(react-storage): refactor Title 

### PR DESCRIPTION
**Pull request by joon-won**
#### Issue #, if available
- Add Title composable component and corresponding control (TitleControl), hook (useTitle)
- Add corresponding unit tests for each of them

#### Checklist

- [x] Have read the [Pull Request Guidelines](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md)
- [x] PR description included
- [x] `yarn test` passes and tests are updated/added
- [ ] PR title and commit messages follow [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) syntax
- [ ] _If this change should result in a version bump_, [changeset added](https://github.com/changesets/changesets/blob/main/docs/adding-a-changeset.md) (This can be done after creating the PR.) This does not apply to changes made to `docs`, `e2e`, `examples`, or other private packages.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/react-storage/src/components/StorageBrowser/controls/TitleControl.tsx [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  className,
+}: ControlProps): React.JSX.Element | null => {
+  const { props } = useTitle();
+  const ResolvedTitle = useResolvedComposable(Title, 'Title');
+
+  if (!props) {
+    return null;
+  }
+
+  const { children, titleClassName } = props;
```
Comment (cshfang) [inline packages/react-storage/src/components/StorageBrowser/controls/TitleControl.tsx:19]: See this comment: https://github.com/aws-amplify/amplify-ui/pull/5935/files/8636d6d874d30ab47bb7a62f863d21356394163d#r1813176271
```suggestion
```

File: packages/react-storage/src/components/StorageBrowser/controls/TitleControl.tsx [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+  return (
+    <ViewElement className={className ? className : titleClassName}>
+      <ResolvedTitle titleClassName={className ? className : titleClassName}>
+        {children}
+      </ResolvedTitle>
```
Comment (cshfang) [inline packages/react-storage/src/components/StorageBrowser/controls/TitleControl.tsx:25]:
```suggestion
      <ResolvedTitle {...props} />
```

File: packages/react-storage/src/components/StorageBrowser/controls/context.ts [hunk 3]
```diff
@@ -2,7 +2,9 @@ import { createContextUtilities } from '@aws-amplify/ui-react-core';
 import { ControlsContext } from './types';
 
 const defaultValue = {
-  data: {},
+  data: {
+    title: 'StorageBrowser',
```
Comment (cshfang) [inline packages/react-storage/src/components/StorageBrowser/controls/context.ts:6]: Hmm I think it's debatable whether or not there should be a default title but I think we should just leave it out for now

File: packages/react-storage/src/components/StorageBrowser/controls/hooks/useTitle.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import { TitleProps } from '../../composables/Title';
+import { useControlsContext } from '../../controls/context';
+import { CLASS_BASE } from '../../views/constants';
+
+const BLOCK_NAME = `${CLASS_BASE}__title`;
```
Comment (cshfang) [inline packages/react-storage/src/components/StorageBrowser/controls/hooks/useTitle.ts:5]: The control should only accept a className for its wrapping div
```suggestion
```

File: packages/react-storage/src/components/StorageBrowser/controls/hooks/useTitle.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  props?: TitleProps;
+};
+
+export const useTitle: UseTitle = () => {
+  const { data } = useControlsContext();
+  const { title } = data;
+
+  return {
+    props: {
+      children: data?.title ? title : '',
```
Comment (cshfang) [inline packages/react-storage/src/components/StorageBrowser/controls/hooks/useTitle.ts:17]: Nit: I think this should be safe?
```suggestion
      children: data?.title,
```

File: packages/react-storage/src/components/StorageBrowser/composables/Title.tsx [hunk 6]
```diff
@@ -0,0 +1,15 @@
+import React from 'react';
+
+import { HeadingElement } from '../context/elements/definitions';
+
+export interface TitleProps {
+  children?: React.ReactNode;
```
Comment (cshfang) [inline packages/react-storage/src/components/StorageBrowser/composables/Title.tsx:6]: Afaik, we'll only ever be providing a string to the Title composables. I would suggest that we just do that here too.
```suggestion
  title?: string;
```

File: packages/react-storage/src/components/StorageBrowser/controls/hooks/useTitle.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+      titleClassName: BLOCK_NAME,
```
Comment (cshfang) [inline packages/react-storage/src/components/StorageBrowser/controls/hooks/useTitle.ts:18]:
```suggestion
```

**Review by cshfang — CHANGES_REQUESTED** [review]

File: packages/react-storage/src/components/StorageBrowser/composables/Title.tsx [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+  titleClassName?: string;
+}
+
+export const Title = ({
+  children,
+  titleClassName,
+}: TitleProps): React.JSX.Element => (
+  <HeadingElement className={titleClassName}>{children}</HeadingElement>
```
[the remark this answers is not in the document]
Comment (joon-won) [reply packages/react-storage/src/components/StorageBrowser/composables/Title.tsx:14]: I see we want the components to share it, got it 👀

File: packages/react-storage/src/components/StorageBrowser/controls/TitleControl.tsx (hunk 1 above)
Comment (joon-won) [reply packages/react-storage/src/components/StorageBrowser/controls/TitleControl.tsx:19]: TIL 👀 , curious what became possible with TS though

File: packages/react-storage/src/components/StorageBrowser/composables/Title.tsx (hunk 6 above)
Comment (joon-won) [reply packages/react-storage/src/components/StorageBrowser/composables/Title.tsx:6]: Got it, for this part I've stuck to the `React.ReactNode` thinking that was a team decision

File: packages/react-storage/src/components/StorageBrowser/controls/hooks/useTitle.ts (hunk 5 above)
Comment (joon-won) [reply packages/react-storage/src/components/StorageBrowser/controls/hooks/useTitle.ts:17]: I see, also removed `{ title } ` above as it's not needed if we stick to it

**Comment by joon-won** [comment]
Integ and Cleaning PR #5958

**Review by ashwinkumar6 — COMMENTED** [review]
Thx Joon, looking good added a few comments.

I think we'll also need to (good to do in another PR)
- delete [TitleControl](https://github.com/aws-amplify/amplify-ui/blob/feat-storage-browser/main/packages/react-storage/src/components/StorageBrowser/views/Controls/Title.tsx#L12) since this acts as a drop in replacement
  - update imports in `StorageBrowser/views/Controls`
- Update views to use new composable title instead of TitleControl

File: packages/react-storage/src/components/StorageBrowser/controls/hooks/useTitle.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
+import { TitleProps } from '../../composables/Title';
+import { useControlsContext } from '../../controls/context';
+
+export type UseTitle = () => {
+  props?: TitleProps;
```
[the remark this answers is not in the document]
Comment (joon-won) [reply packages/react-storage/src/components/StorageBrowser/controls/hooks/useTitle.ts:5]: Oh got it, thanks for sharing that. I'll match the style

File: packages/react-storage/src/components/StorageBrowser/controls/types.ts [hunk 10]
```diff
@@ -39,12 +39,14 @@ interface Location {
 
 export interface ControlsContext {
   data: {
+    isDataRefreshDisabled?: boolean;
```
Comment (Samaritan1011001) [inline packages/react-storage/src/components/StorageBrowser/controls/types.ts:42]: This is already present on line 45.

File: packages/react-storage/src/components/StorageBrowser/composables/types.ts [hunk 11]
```diff
@@ -12,6 +15,8 @@ export interface Composables {
   DropZone: React.ComponentType<DropZoneProps>;
   Navigation: React.ComponentType<NavigationProps>;
   StatusDisplay: React.ComponentType<StatusDisplayProps>;
+  ActionStart: React.ComponentType<ActionStartProps>;
```
Comment (Samaritan1011001) [inline packages/react-storage/src/components/StorageBrowser/composables/types.ts:18]: This is also not needed I think

**Review by Samaritan1011001 — APPROVED** [review]

**Comment by joon-won** [comment]
Please refer to new PR with rebase https://github.com/aws-amplify/amplify-ui/pull/6013

**Pull request closed without merging**